### PR TITLE
[InstCombine] Improve folding of `icmp pred (and X, Mask/~Mask), Y)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for expanding `foldICmpWithLowBitMaskedVal`; NFC
- [InstCombine] Improve mask detection in `foldICmpWithLowBitMaskedVal`
- [InstCombine] Recognize `(icmp eq/ne (and X, ~Mask), 0)` pattern in `foldICmpWithLowBitMaskedVal`

Proofs for new patterns: https://alive2.llvm.org/ce/z/fSQ3nZ

### CODE REVIEW

**Review by AZero13 — APPROVED** [review]
These changes look great!

File: llvm/include/llvm/IR/PatternMatch.h
Comment (dtcxzyw) [inline llvm/include/llvm/IR/PatternMatch.h:567]: Could you add some unit tests in `llvm/unittests/IR/PatternMatch.cpp`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4071]:
```suggestion
// Returns whether V is a Mask ((X + 1) & X == 0) or ~Mask (-Pow2OrZero)
```

File: llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll:4]: @nikic Do we still need to declare these intrinsic functions explicitly?

See also https://github.com/llvm/llvm-project/pull/78251.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4082]: Why not move the logic into the following switch statement?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4086]: Could you please provide the alive2 proofs?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4103]: Is it a Mask or a ~Mask?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4103]: Mask: its the [blsmsk](https://www.felixcloutier.com/x86/blsmsk) pattern from BMI2 extension.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4086]: Sorry, in the commit no the summary: https://alive2.llvm.org/ce/z/fSQ3nZ
added to summary as well.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4104]:
```suggestion
    return  !Not && match(V, m_c_Xor(m_Value(X), m_Add(m_Deferred(X), m_AllOnes())));
```

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4104]: Ahh, good catch. Adding a test.

File: llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll:4]: Don't seem to need it. Dropping in v2

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4078]: What is the compile-time impact of this patch?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: Adding a patch. Will post seperately shortly.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: see: https://github.com/llvm/llvm-project/pull/81576

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4078]: No meaningful affect: https://llvm-compile-time-tracker.com/compare.php?from=fb48fd18c240574841378defacadff34238089bb&to=ec40daf22da38fb28165ea1e757493470fec218e&stat=instructions:u

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4113]: Ill post v3 after this gets in.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4297]: Use `IC.getFreelyInverted`?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4297]: If its freely invertable `(not M)` will simplify later on. Seems simpler to let the existing pipeline clean it up. No?

**Comment by goldsteinn** [comment]
ping

**Comment by goldsteinn** [comment]
ping

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Comment by goldsteinn** [comment]
rebased

**Comment by goldsteinn** [comment]
ping @nikic

**Comment by goldsteinn** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4211]: Can you instead call this fold from foldICmpCommutative() and only handle one direction here?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4082]: I dont think its an issue for llvm, but on systems where `nullptr` is non-zero, does `!I` still work?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4211]: Done, commit to do so is split into NFC.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4186]: They seem about the same to me, happy to change.

**Pull request closed without merging**
